File: src/utils/whatsappApi.js
import { auth } from '../firebase/config';

// Calls the serverless WhatsApp send endpoint (api/whatsapp/send.js).
// Sends via the Meta Cloud API using pre-approved templates.
//
//   type      'renewal' | 'payment' | 'class' | 'announcement'
//   memberIds array of member doc ids in gyms/{gymId}/members
//   extra     optional { body, className, amount }
//
// NOTE: this only works on the deployed Vercel site (or via `vercel dev`).
// Under plain `npm run dev` (Vite) the /api function isn't served.
export async function sendWhatsApp({ gymId, type, memberIds, extra }) {
  const user = auth.currentUser;
  if (!user) throw new Error('You must be signed in');
  const token = await user.getIdToken();

  const res = await fetch('/api/whatsapp/send', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({ gymId, type, memberIds, extra }),
  });

  let data;
  try { data = await res.json(); } catch { data = {}; }
  if (!res.ok) throw new Error(data.error || `Send failed (HTTP ${res.status})`);
  return data; // { sent, failed, results }
}

File: mobile-app/lib/services/whatsapp_api.dart
import 'dart:convert';

import 'package:firebase_auth/firebase_auth.dart';
import 'package:http/http.dart' as http;

/// Calls the deployed WhatsApp Cloud API send endpoint — mirrors the web
/// `src/utils/whatsappApi.js`. Used ONLY for Utility templates (renewal +
/// payment); Marketing messages (class reminders, announcements) stay on free
/// wa.me links via actions.dart `openWhatsApp`.
const String _endpoint = 'https://app-kilos.devlofttech.com/api/whatsapp/send';

class WhatsAppApiResult {
  final int sent;
  final int failed;
  final String? error;
  const WhatsAppApiResult({this.sent = 0, this.failed = 0, this.error});
  bool get ok => error == null && failed == 0;
}

/// POST { gymId, type, memberIds, extra } with the caller's Firebase ID token.
/// type: 'renewal' | 'payment'
Future<WhatsAppApiResult> sendWhatsAppApi({
  required String gymId,
  required String type,
  required List<String> memberIds,
  Map<String, dynamic>? extra,
}) async {
  final user = FirebaseAuth.instance.currentUser;
  if (user == null) {
    return WhatsAppApiResult(failed: memberIds.length, error: 'You must be signed in');
  }

  try {
    final token = await user.getIdToken();
    final res = await http.post(
      Uri.parse(_endpoint),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token',
      },
      body: jsonEncode({
        'gymId': gymId,
        'type': type,
        'memberIds': memberIds,
        if (extra != null) 'extra': extra,
      }),
    );

    Map<String, dynamic> data = {};
    try {
      data = jsonDecode(res.body) as Map<String, dynamic>;
    } catch (_) {/* non-JSON body */}

    if (res.statusCode != 200) {
      return WhatsAppApiResult(
        failed: memberIds.length,
        error: data['error']?.toString() ?? 'Send failed (${res.statusCode})',
      );
    }
    return WhatsAppApiResult(
      sent: (data['sent'] as num?)?.toInt() ?? 0,
      failed: (data['failed'] as num?)?.toInt() ?? 0,
    );
  } catch (e) {
    return WhatsAppApiResult(failed: memberIds.length, error: e.toString());
  }
}
